app/scripts/background.js

```diff
@@ -330,9 +330,9 @@ function setupController(initState, initLangCode) {
 
       if (processName === ENVIRONMENT_TYPE_POPUP) {
         popupIsOpen = true;
-
         endOfStream(portStream, () => {
           popupIsOpen = false;
+          controller.onPopUpClose();
```

Comment:
I just a realized a scenario where this doesn't work:

- have a transaction open in a full tab and also have the popup open. Gas will be polling
- close the popup, but leave the full tab with the transaction open. Gas _should_ continue polling, but won't, because we just closed the popup

Reply:
:thinking:

Reply:
First, as an improvement, instead of calling `onPopupClose` in the setter, lets call it in each of the `endOfStream` callbacks in this file. (Also, see my above comment on the rename of `onPopupClose`). We can do this after `controller.isClientOpen` is set, and only do it if it is false.

I briefly commented on this when we are talking yesterday, but with a bit more sleep in me, I think I have come down more firmly on the side of not wanting to have unexpected sideeffects in a setter. If we are overwriting a setter method, I don't think it should do much more than setting. That could be unexpected for a reader of `background.js`

Reply:
So that will safely handle the case where all instances of client connections are closed.

Reply:
For the edge in my first comment in this thread, or the other edge cases we discussed yesterday when you have multiple tabs or notification windows open and close some but not all, we need something more.

I think what we have to do is:
- maintain a record of poll tokens per environment (e.g. `popupGasPollTokens`, `notificationGasPollTokens`, `fullScreenGasPollTokens`)
- in the `endOfStream` callback for each environment, if it is not the case where `controller.isClientOpen === false`, then just disconnect polling for each token from that environment

We could track those records in the `appState` controller. We can add / remove them in any place in the frontend where we get a gas poll token upon starting gas polling.

Reply:
Then I think we will need separate `onPopupEnvClose()`, `onNotificationEnvClose()` and `onFullScreenEnvClose()` methods for each of the three environments. The method we call when all are closed could then call each of these methods.

Reply:
> First, as an improvement, instead of calling `onPopupClose` in the setter, lets call it in each of the `endOfStream` callbacks in this file. 

@danjm Would this include `processName === ENVIRONMENT_TYPE_NOTIFICATION`? No right?